wallet/udb/txmined.go

```diff
@@ -321,6 +321,88 @@ func (s *Store) UpdateProcessedTxsBlockMarker(dbtx walletdb.ReadWriteTx, hash *c
 	return nil
 }
 
+// SetBirthblock sets a birthday block and hash. It also sets the birthday time
+// should be set bool to false.
+//
+// [0:32]  Block header hash (32 bytes)
+// [32:36] Birthblock block height (4 bytes)
+func (s *Store) SetBirthblock(dbtx walletdb.ReadWriteTx, blockHash *chainhash.Hash, height uint32) error {
+	ns := dbtx.ReadWriteBucket(wtxmgrBucketKey)
+	err := s.setBirthblock(ns, blockHash, height)
+	if err != nil {
+		return err
+	}
+	return s.setBirthdayTimeBool(ns, false)
+}
+
+func (s *Store) setBirthblock(ns walletdb.ReadWriteBucket, blockHash *chainhash.Hash, height uint32) error {
+	v := make([]byte, chainhash.HashSize+4)
+	copy(v[:], blockHash[:])
+	byteOrder.PutUint32(v[chainhash.HashSize:], uint32(height))
+	return ns.Put(rootBirthblock, v)
+}
+
+// Birthblock returns the currently set wallet birthday block hash and height. If
+// not set a nil hash is returned.
+func (s *Store) Birthblock(dbtx walletdb.ReadTx) (*chainhash.Hash, uint32) {
+	ns := dbtx.ReadBucket(wtxmgrBucketKey)
+	v := ns.Get(rootBirthblock)
+	if len(v) != chainhash.HashSize+4 {
+		return nil, 0
+	}
+	var h chainhash.Hash
+	copy(h[:], v[:chainhash.HashSize])
+	return &h, byteOrder.Uint32(v[chainhash.HashSize:])
+}
+
+// SetBirthdayTime sets a time that the birthday should be set from. Also sets
+// the bool that decides if the birthday should be set to true.
+//
+// [0:1] should set bool (1 byte)
+// [1:9] Birthday time (8 bytes)
+func (s *Store) SetBirthdayTime(dbtx walletdb.ReadWriteTx, birthday *time.Time) error {
+	ns := dbtx.ReadWriteBucket(wtxmgrBucketKey)
+	v := make([]byte, 1+8)
+	v[0] = 1 // true
+	byteOrder.PutUint64(v[1:], uint64(birthday.Unix()))
+	return ns.Put(rootBirthdayTime, v)
+}
+
+// SetBirthdayTimeBool sets only the bool that decides if the birthday should
+// be set. Will error if the birthday time was never set.
+func (s *Store) SetBirthdayTimeBool(dbtx walletdb.ReadWriteTx, shouldSet bool) error {
```

Comment:
I'd rename this to "NeedsBirthdayBlock" to make it clearer what this bool refers to. 

Alternatively, you could check whether the birthblock hash is set to determine whether you need to set it or not (if it's nil, you need to set it). Then you remove the need for this separate bit.

Personally, I'd unify all the birthday state into a single struct:

```go
type BirthdayState struct {
  Hash *chainhash.Hash
  Height uint32
  Time time.Time
}
```

You can use different functions to modify it in the DB (so that assumptions about it hold), but then it's a single call to fetch all birthday related information.

Reply:
> Alternatively, you could check whether the birthblock hash is set to determine whether you need to set it or not (if it's nil, you need to set it). Then you remove the need for this separate bit.

If the user restores and does not set a birthday, it will not have one, but I guess we can just save the genesis hash then, it will be the same.